notifier/notify_windows.go

```diff
@@ -2,351 +2,342 @@ package main
-}
+	new.addButton, err = new.dll.FindProc("PortmasterToastAddButton")
+	if err != nil {
+		return fmt.Errorf("PortmasterToastAddButton not found %q", err)
+	}
 
-// internal functions:
+	new.setImage, err = new.dll.FindProc("PortmasterToastSetImage")
+	if err != nil {
+		return fmt.Errorf("PortmasterToastSetImage not found %q", err)
+	}
 
-func initNotifierPath() {
-	notifierPathMutex.Lock()
-	defer notifierPathMutex.Unlock()
+	new.showNotification, err = new.dll.FindProc("PortmasterToastShow")
+	if err != nil {
+		return fmt.Errorf("PortmasterToastShow not found %q", err)
+	}
 
-	if notifierPath == "" {
-		var err error
-		notifierPath, err = getPath("notifier-snoretoast")
+	new.setActivatedCallback, err = new.dll.FindProc("PortmasterToastActivatedCallback")
+	if err != nil {
+		return fmt.Errorf("PortmasterActivatedCallback not found %q", err)
+	}
 
-		if err != nil {
-			log.Errorf("failed obtaining SnoreToast-Path: %s %s", err, notifierPath)
-			return
-		}
+	new.setDismissedCallback, err = new.dll.FindProc("PortmasterToastDismissedCallback")
+	if err != nil {
+		return fmt.Errorf("PortmasterToastDismissedCallback not found %q", err)
 	}
-}
 
-func notificationListener() {
-	log.Debugf("starting Callback-Pipe for SnoreToast: %s", pipeName)
-	l, err := winio.ListenPipe(pipeName, nil)
+	new.setFailedCallback, err = new.dll.FindProc("PortmasterToastFailedCallback")
 	if err != nil {
-		log.Errorf("failed to start Pipe for SnoreToast: %s", err)
-		return
+		return fmt.Errorf("PortmasterToastFailedCallback not found %q", err)
 	}
-	defer l.Close()
 
-	for {
-		conn, err := l.Accept()
-		if err != nil {
-			log.Errorf("failed to accept namedPipe-connection (for receiving Callbacks from SnoreToast): %s", err)
-			continue
-		}
-		go handlePipeMessage(conn)
+	new.hideNotification, err = new.dll.FindProc("PortmasterToastHide")
+	if err != nil {
+		return fmt.Errorf("PortmasterToastHide not found %q", err)
 	}
+
+	lib = new
+	return nil
 }
 
-func handlePipeMessage(conn net.Conn) {
-	data := parseSnoreToastActionCallback(readWideStringAsUnicode(conn))
-	log.Debugf("handling PipeMessage: %+v", data)
+// API called functions:
 
-	if data.action == "timedout" {
+// Show shows the notification.
+func (n *Notification) Show() {
+	if lib == nil {
+		log.Error("notify: library not properly loaded")
 		return
 	}
 
-	notificationsLock.Lock()
-	var n *Notification
-	for _, elem := range notifications {
-		if elem.GUID == data.id {
-			n = elem
-			break
-		}
-	}
-	notificationsLock.Unlock()
-	if n == nil {
-		return
+	if !libInitialized {
+		log.Error("notify: not initialized")
 	}
+
+	// Lock notification
 	n.Lock()
 	defer n.Unlock()
 
-	switch data.action {
-	case "dismissed": // do nothing
-	case "clicked": // like dismissed: do nothing
-	case "buttonClicked":
-		for _, actions := range n.AvailableActions {
-			if actions.Text == data.button {
-				n.SelectAction(actions.ID)
-				return
-			}
+	// Lock library
+	lib.Lock()
+	defer lib.Unlock()
```

Comment:
I've made bad experiences with double locking. Is there an easy way to avoid this?